src/lib/FreeC/Pass/CompletePatternPass.hs

```diff
@@ -0,0 +1,154 @@
+where
+
+import           Control.Monad                  ( unless )
+
+import           FreeC.Environment
+import           FreeC.Environment.Entry
+import qualified FreeC.IR.Syntax               as IR
+import           FreeC.IR.SrcSpan
+import           FreeC.Monad.Converter
+import           FreeC.Monad.Reporter
+import           FreeC.Pass
+import           FreeC.Pretty                   ( showPretty )
+
+-- | Checks that all functions of a given module have complete pattern matching.





























































+--







+--   The pattern matching is complete if there is exactly one case alternative





























































+--   for each constructor of the corresponding type.





























































+completePatternPass :: Pass IR.Module
+completePatternPass ast = do
+  mapM_ checkPatternFuncDecl (IR.modFuncDecls ast)
+  return ast
+
+-- | Checks a function declaration for incomplete patterns





















































.







+checkPatternFuncDecl :: IR.FuncDecl -> Converter ()
+checkPatternFuncDecl funcDecl = checkPatternExpr (IR.funcDeclRhs funcDecl)
+ where
+  -- | Checks an expression for incomplete patterns.







+  checkPatternExpr :: IR.Expr -> Converter ()
+  checkPatternExpr (IR.Case srcSpan exprScrutinee exprAlts _) = do
+    checkPatternExpr exprScrutinee
+    mapM_ (checkPatternExpr . IR.altRhs) exprAlts
+    case IR.exprType exprScrutinee >>= getTypeConName of
+      Nothing       -> failedPatternCheck srcSpan
+      Just typeName -> do
+        maybeEntry <- inEnv $ lookupEntry IR.TypeScope typeName
+        let altConNames = map (IR.conPatName . IR.altConPat) exprAlts
+        case maybeEntry of
+          Just entry -> unfoldTypeSyn entry altConNames srcSpan
+          _          -> failedPatternCheck srcSpan
+  checkPatternExpr (IR.App _ lhr rhs _) =
+    checkPatternExpr lhr >> checkPatternExpr rhs
+  checkPatternExpr (IR.TypeAppExpr _ lhr _ _) = checkPatternExpr lhr
+  checkPatternExpr (IR.If _ exprCond exprThen exprElse _) =
+    checkPatternExpr exprCond
+      >> checkPatternExpr exprThen
+      >> checkPatternExpr exprElse
+  checkPatternExpr (IR.Lambda _ _ lambdaRhs _) = checkPatternExpr lambdaRhs
+  checkPatternExpr IR.Con{}                    = return ()
+  checkPatternExpr IR.Var{}                    = return ()
+  checkPatternExpr IR.Undefined{}              = return ()
+  checkPatternExpr IR.ErrorExpr{}              = return ()
+  checkPatternExpr IR.IntLiteral{}             = return ()
+
+  -- | converts type synonyms and performs check on data entries






+  unfoldTypeSyn :: EnvEntry -> [IR.ConName] -> SrcSpan -> Converter ()
+  unfoldTypeSyn entry altConNames srcSpan | isDataEntry entry = do
+    performCheck (entryConsNames entry) altConNames srcSpan
+  unfoldTypeSyn entry altConNames srcSpan | isTypeSynEntry entry = do
+    case getTypeConName (entryTypeSyn entry) of
```

Comment:
Consider the following type synonym.

```haskell
type Id a = a
```

The pattern matching in the following `case` expression is complete, but no `DataEntry` will be found by `unfoldTypeSyn`.

```haskell
case (x :: Id [a]) of
  []     -> …
  x : xs -> …
```

Try to reuse [`expandAllTypeSynonyms`](https://freeproving.github.io/free-compiler/docs/master/FreeC-IR-TypeSynExpansion.html#v:expandAllTypeSynonyms).

Reply:
Oh, I searched for something similar to `expandAllTypeSynonyms` but I couldn't find it.